core/src/Kokkos_OpenACCSpace.hpp

```diff
@@ -0,0 +1,266 @@
+      Kokkos::Experimental::OpenACCSpace, void>;
+};
+}  // namespace Experimental
+}  // namespace Kokkos
+
+namespace Kokkos {
+namespace Impl {
+
+template <>
+class SharedAllocationRecord<Kokkos::Experimental::OpenACCSpace, void>
+    : public HostInaccessibleSharedAllocationRecordCommon<
+          Kokkos::Experimental::OpenACCSpace> {
+ private:
+  friend class HostInaccessibleSharedAllocationRecordCommon<
+      Kokkos::Experimental::OpenACCSpace>;
+  friend class SharedAllocationRecordCommon<Kokkos::Experimental::OpenACCSpace>;
+  friend Kokkos::Experimental::OpenACCSpace;
+
+  using base_t = HostInaccessibleSharedAllocationRecordCommon<
+      Kokkos::Experimental::OpenACCSpace>;
+  using RecordBase = SharedAllocationRecord<void, void>;
+
+  SharedAllocationRecord(const SharedAllocationRecord&) = delete;
+  SharedAllocationRecord& operator=(const SharedAllocationRecord&) = delete;
+
+  /**\brief  Root record for tracked allocations from this OpenACCSpace
+   * instance */
+  static RecordBase s_root_record;
+
+  const Kokkos::Experimental::OpenACCSpace m_space;
+
+ protected:
+  ~SharedAllocationRecord();
+  SharedAllocationRecord() = default;
+
+  template <typename ExecutionSpace>
+  SharedAllocationRecord(
+      const ExecutionSpace& /*exec_space*/,
+      const Kokkos::Experimental::OpenACCSpace& arg_space,
+      const std::string& arg_label, const size_t arg_alloc_size,
+      const RecordBase::function_type arg_dealloc = &deallocate)
+      : SharedAllocationRecord(arg_space, arg_label, arg_alloc_size,
+                               arg_dealloc) {}
+
+  SharedAllocationRecord(
+      const Kokkos::Experimental::OpenACCSpace& arg_space,
+      const std::string& arg_label, const size_t arg_alloc_size,
+      const RecordBase::function_type arg_dealloc = &deallocate);
+
+ public:
+  KOKKOS_INLINE_FUNCTION static SharedAllocationRecord* allocate(
+      const Kokkos::Experimental::OpenACCSpace& arg_space,
+      const std::string& arg_label, const size_t arg_alloc_size) {
+    if (acc_on_device(acc_device_host)) {
+      return new SharedAllocationRecord(arg_space, arg_label, arg_alloc_size);
+    } else {
+      return nullptr;
+    }
+  }
+};
+
+}  // namespace Impl
+}  // namespace Kokkos
+
+//----------------------------------------------------------------------------
+//----------------------------------------------------------------------------
+
+namespace Kokkos {
+namespace Impl {
+
+// FIXME_OPENACC: implement all possible deep_copies
+template <class ExecutionSpace>
+struct DeepCopy<Kokkos::Experimental::OpenACCSpace,
+                Kokkos::Experimental::OpenACCSpace, ExecutionSpace> {
+  DeepCopy(void* dst, const void* src, size_t n) {
+    if (n > 0) acc_memcpy_device(dst, const_cast<void*>(src), n);
+  }
+  DeepCopy(const ExecutionSpace& exec, void* dst, const void* src, size_t n) {
+    exec.fence();
+    if (n > 0) acc_memcpy_device(dst, const_cast<void*>(src), n);
```

Comment:
Maybe, `multiple execution space instances` is a right term to use? 
What I meant by  is that the current OpenACC backend implementation assumes that there is only one device execution space instance, and all the device operations (e.g., memory transfers, kernel launches, etc.) are implemented to be synchronous, which does not violate the Kokkos execution semantics with the single execution space instance.
(What I assume is that the main purpose of supporting multiple execution space instances is to enable asynchronous/concurrent device operations with each other, and thus with a single execution space instance support, it is OK to execute device operations synchronously. If multiple execution space instances are allowed, asynchronous device operations will be necessary to enable concurrent executions.)

Reply:
It does not violate semantics sure, but it is a performance bug.  That said I am fine with fixing later.

Reply:
yes and no, if you have for example multiple instances but each is owned by its own std::thread then it doesn't need to be asynchronous, but it needs to be thread safe (i.e. internal state is per instance and not like static variables).

Reply:
Can any of you point me to documents where I can find information on where global synchronizations are expected and where execution space instance specific synchronizations are expected?
Regarding the performance bug mentioned by @dalg24, I almost finished implementing the multiple instance support in the OpenACC backend, but when I tested the performance with some local test programs, their performance was worse than the single instance version, mainly because of too many synchronizations. Therefore, it will be very helpful for me to know the minimum synchronization points to reduce unnecessary synchronization overheads.

Reply:
Best documentation of the most common cases is in the lectures: https://www.youtube.com/watch?v=xEAyOod57-c&feature=youtu.be&t=2949